cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/AuthenticationServiceMain.java

```diff
@@ -0,0 +1,152 @@
+ */
+public class AuthenticationServiceMain extends AbstractServiceMain<EnvironmentOptions> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuthenticationServiceMain.class);
+
+  public static void main(String[] args) throws Exception {
+    main(AuthenticationServiceMain.class, args);
+  }
+
+  @Override
+  protected List<Module> getServiceModules(
+      MasterEnvironment masterEnv, EnvironmentOptions options, CConfiguration cConf) {
+    return Arrays.asList(
+        getDataFabricModule(),
+        new ZKClientModule(),
+        new SecurityModules().getDistributedModules(),
+        new MessagingClientModule());
+  }
+
+  @Override
+  protected void addServices(
+      Injector injector, List<? super Service> services, List<? super AutoCloseable> closeableResources,
+      MasterEnvironment masterEnv, MasterEnvironmentContext masterEnvContext, EnvironmentOptions options) {
+
+    MessagingService messagingService = injector.getInstance(MessagingService.class);
+    if (messagingService instanceof Service) {
+      services.add((Service) messagingService);
+    }
+
+    CConfiguration configuration = injector.getInstance(CConfiguration.class);
+
+    if (configuration.getBoolean(Constants.Security.ENABLED)) {
+
+      services.add(new AbstractService() {
+
+        @Override
+        protected void doStart() {
+
+          ZKClientService zkClientService = injector.getInstance(ZKClientService.class);
+          ExternalAuthenticationServer authServer = injector.getInstance(ExternalAuthenticationServer.class);
+
+          try {
+            LOG.info("Starting AuthenticationServer.");
+
+            // Enable Kerberos login
+            SecurityUtil.enableKerberosLogin(configuration);
+
+            io.cdap.cdap.common.service.Services.startAndWait(
+                zkClientService,
+                configuration.getLong(
+                    Constants.Zookeeper.CLIENT_STARTUP_TIMEOUT_MILLIS),
+                TimeUnit.MILLISECONDS,
+                String.format(
+                    "Connection timed out while trying to start " +
+                        "ZooKeeper client. Please verify that the " +
+                        "ZooKeeper quorum settings are correct in " +
+                        "cdap-site.xml. Currently configured as: %s",
+                    configuration.get(Constants.Zookeeper.QUORUM)));
+            authServer.startAndWait();
+
+            notifyStarted();
+          } catch (Exception e) {
+            Throwable rootCause = Throwables.getRootCause(e);
+            if (rootCause instanceof ServiceBindException) {
+              LOG.error("Failed to start Authentication Server: {}", rootCause.getMessage());
+            } else {
+              LOG.error("Failed to start Authentication Server", e);
+            }
+          }
+        }
+
+        @Override
+        protected void doStop() {
+
+          ZKClientService zkClientService = injector.getInstance(ZKClientService.class);
```

Comment:
It's not good to assume that the ZKClientService is a singleton binding.

Reply:
ZKClientModule defines it as a singleton. I do understand that it is done in separate place.

Injector does not provide a way to access all instances - I have been researching in that space. Is the code bellow is the way you are thinking all instances should be retrieved?

 ```
      for (Key<?> key : injector.getAllBindings().keySet()) {
           if (YourInterface.class.isAssignableFrom(key.getTypeLiteral().getRawType())) {
            YourInterface yourInterface = (YourInterface) injector.getInstance(key);
            allYourInterfaces.add(yourInterface);
       }
```

isAssignableFrom method matches all subclasses - in this situation we could just use `==`

Source: https://stackoverflow.com/questions/6085200/how-to-get-all-implementors-subclasses-of-an-interface-with-guice/40558788
     
Is that how you see it?

Reply:
What I meant was instead of relying on the binding being singleton,  it is better to just get the instance once and hold the instance in a field so that it is for sure it is the same instance.

Reply:
Thanks, makes sense. I have updated code.